### DISCUSSION TITLE
Simple embind example doesn't seem to work

### DISCUSSION THREAD

**Discussion by k1nder10** [emscripten-core/emscripten#17246] [Show and tell]
main.cpp
```
#include <emscripten/bind.h>

struct Point {
  int x;
  int y;
};

EMSCRIPTEN_BINDINGS(MainModule) {
  emscripten::value_object<Point>("Point")
  .field("x", &Point::x)
  .field("y", &Point::y)
  ;
}

int main() {}
```

my_js.js
```
var Module = require('./main.js')

Module.onRuntimeInitialized = async function(){
  var point = new Module.Point();
}
```

Build command: emcc main.cpp -o main.js -lembind
emsdk version is 3.1.13

Runtime error:
TypeError: Module.Point is not a constructor
    at Object.Module.onRuntimeInitialized (C:\projects\opengl_wasm\bind_small_example\my_js.js:7:15)
    at doRun (C:\projects\opengl_wasm\bind_small_example\main.js:3138:71)
    at run (C:\projects\opengl_wasm\bind_small_example\main.js:3155:5)
    at runCaller (C:\projects\opengl_wasm\bind_small_example\main.js:3060:19)
    at removeRunDependency (C:\projects\opengl_wasm\bind_small_example\main.js:1404:7)
    at receiveInstance (C:\projects\opengl_wasm\bind_small_example\main.js:1592:5)
    at receiveInstantiationResult (C:\projects\opengl_wasm\bind_small_example\main.js:1610:5)

**Comment by k1nder10** [comment]
I tried to expose Point as emscripten::class and it works fine. But neither emscripten::value_object nor emscripten::value_array works
```
class_<Point>("Point")
  .constructor<>();
```

**Comment by brendandahl** [comment]
I don't think value_object is meant to be used with `new`. There are some examples using value_object in the tests. It uses a function to get one:
https://github.com/emscripten-core/emscripten/blob/df69d563c2aebb6b86f709ae86a61f4dbc1bf370/tests/embind/embind_test.cpp#L2579-L2582

https://github.com/emscripten-core/emscripten/blob/df69d563c2aebb6b86f709ae86a61f4dbc1bf370/tests/embind/embind_test.cpp#L2588-L2590

https://github.com/emscripten-core/emscripten/blob/df69d563c2aebb6b86f709ae86a61f4dbc1bf370/tests/embind/embind.test.js#L2408-L2412

**Reply by k1nder10** [reply]
Okay, thanks.
A couple more questions for you. I want the following functions to be implemented in JS. I use --js-library for this.
```
// in C++:
struct Point { 
  int x;
  int y;
};

Point create_point() { // in the emscripten_bindings 
  return {};
}

extern "C" extern void fill_point(Point* p);
extern "C" extern Point get_point();

// in JS's mergeInto:
fill_point: function(point) {
  ??? // 1. How to implement this? point is a pointer!
}

get_point: function() {
  let p = Module.create_point();
  p.x = 5;
  p.y = 10;
  return p; // 2. When call this functions from C++ it returns empty Point, not filled one.
}

// 3. Is it okay that Point is not under extern "C" block?
```

**Reply by k1nder10** [reply]
As I understand I cannot implement either of them. The only way to do so is to wrap them in C++. Is that true?

```
struct Point {
  int x;
  int y;
};

extern "C" int get_x(); // defined in JS and uses only fundamental types which is OK
extern "C" int get_y(); // same as get_x

void fill_point(Point* p) {
  p->x = get_x();
  p->y = get_y();
}

Point get_point() {
  Point p;
  p.x = get_x();
  p.y = get_y();
}
```

**Reply by k1nder10** [reply]
@sbc100 Can you take a look at this? Just want to make sure that I understand it properly

**Reply by sbc100** [reply]
I'm afraid I'm not much of an embind expert, but its seems like the world of JS library functions and the world of embind are separate.    As far as I know JS library function only take basic types (and pointers)..   So I don't think you can make a JS library function that takes a Point object directly... but I could be wrong.

**Reply by brendandahl** [reply]
Passing by value isn't really wired up that way to work between JS<->C++ and doesn't really behave how it would in a pure C/C++ program. value_object is more a helper for creating a way to send a non mutable object between JS/C++. If you did want to mix more JS/C++ you could do something like:

```c++
#include <emscripten/bind.h>
#include <emscripten.h>

using namespace emscripten;

// in C++:
struct Point { 
  int x;
  int y;

  int getX() const {
    return x;
  }
  void setX(int aX) {
    x = aX;
  }

  int getY() const {
    return y;
  }
  void setY(int aY) {
    y = aY;
  }

  static Point* load(int ptr) {
    return (Point*) ptr;
  }
};

extern "C" extern void fill_point(Point* p);
extern "C" extern Point* get_point();

int main() {
  Point* p = get_point();
  printf("Point1 x=%d y=%d\n", p->x, p->y);
  fill_point(p);
  printf("Point2 x=%d y=%d\n", p->x, p->y);
  delete p;
  return 0;
}

EMSCRIPTEN_BINDINGS(order) {
    class_<Point>("Point")
        .constructor()
        .property("x", &Point::getX, &Point::setX)
        .property("y", &Point::getY, &Point::setY)
        .class_function("load", &Point::load, allow_raw_pointers());
}
```

```js
mergeInto(LibraryManager.library, {
    get_point: function() {
        let p = new Module.Point();
        p.x = 5;
        p.y = 10;

        return p.$$.ptr;
    },
    fill_point: function(ptr) {
        let p = Module.Point.load(ptr);
        p.x = 4;
        p.y = 2;
    }
  });
```
and this would output:

```
Point1 x=5 y=10
Point2 x=4 y=2
```

There may be a cleaner way to do this, I'm still getting use to embind as well.

**Reply by k1nder10** [reply]
Yeah, it seems like they can't be used together. Though I made up the following example, I don't know whether it should work.
Here we use pointers, which are allowed in JS libraries. However, the code works as if fillPoint was never called. Is it expected?
 ```
 // point.cpp
extern "C" {
struct Point {
  int x;
  int y;
};

extern void fillPoint(Point *p); // to be implemented in JS

Point createPoint() { 
  return {};
}
}

EMSCRIPTEN_BINDINGS(PointModule) {
  emscripten::value_object<Point>("Point")
      .field("x", &Point::x)
      .field("y", &Point::y);

  emscripten::function("createPoint", &createPoint);
}

int main() {
  Point p{1, 2};
  fillPoint(&p);
  std::cout << p.x << ' ' << p.y << std::endl; // prints 1, 2
}
  
 // point.js
mergeInto(LibraryManager.library, {
  fillPoint__sig: 'vp',
  fillPoint: function(point) {
    let p = Module.createPoint();
    p.x = 5;
    p.y = 10;
    point = p;
  }
});

emcc point.cpp --js-library point.js --bind -o point_out.html
 ```

**Reply by k1nder10** [reply]
Ok. typeof() of a C++ pointer says it is a number. Then I understand why it doesn't work. But then the question is how can I write to a C++ pointer in JS? 
```
 fillPoint: function(point) {
   console.log(typeof(point)); // number
   let p = Module.createPoint();
   console.log(typeof(p)); // object
   p.x = 5;
   p.y = 10;
   point = p; // assigning object to number
   // Question: how to treat 'point' variable as object ?? how to dereference 'point' variable?
 }
```

**Reply by k1nder10** [reply]
@brendandahl 
I didn't see your comment while writing those two above. Thank you very much!
